Enable cluster tests on blockfrost network

v0d1ch opened the pull request:
fix #2048 

- Idea was to be able to run our smoke-test using blockfrost network but turned out we could enable all cluster test to be  potentially run on blockfrost more easily.

- Successfull smoke-test run https://github.com/cardano-scaling/hydra/actions/runs/15754166894/job/44405708211
---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Enable cluster tests on blockfrost network

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

noonio commented on .github/workflows/smoke-test.yaml:
```diff
@@ -60,7 +62,13 @@ jobs:
+
+        if [[ ${{secrets.blockfrost_token}} != '' ]]; then
+          echo "${{secrets.blockfrost_token}}" > blockfrost-project.txt
+          nix develop .#exes --command bash -c "hydra-cluster --${{inputs.network}} --state-directory ${state_dir} ${HYDRA_SCRIPTS_ARG} ${USE_MITHRIL_ARG}"
```
Oops; this line doesn't want to be _within_ the `if` statement here!

v0d1ch replied:
:sweat_smile:

noonio commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -82,16 +84,20 @@ bench startingNodeId timeoutSeconds workDir dataset = do
             let hydraTracer = contramap FromHydraNode tracer
-            withHydraCluster hydraTracer workDir nodeSocket startingNodeId cardanoKeys hydraKeys hydraScriptsTxId 10 $ \clients -> do
+
+            let nodeSocket' = case Backend.getOptions backend of
```
Maybe a bit nicer to move this check up a bit, so we don't also publish the scripts if someone accidentally runs it with blockfrost?

noonio commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -1,19 +1,28 @@
 import Test.Hydra.Prelude (withTempDir)
 
+blockfrostProjectPath :: FilePath
+blockfrostProjectPath = "./blockfrost-project.txt"
```
Don't we already have this defined somewhere?

noonio commented on hydra-cluster/src/Hydra/Cluster/Mithril.hs:
```diff
@@ -63,15 +63,24 @@ downloadLatestSnapshotTo tracer network directory = do
     Preproduction -> "https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-preprod/genesis.vkey"
     Preview -> "https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/pre-release-preview/genesis.vkey"
     Sanchonet -> "https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/testing-sanchonet/genesis.vkey"
+    BlockfrostPreview -> ""
```
Probably at least worth an XXX/TODO style comment here; i.e. it would be nicer to convert this to a maybe type somewhere? But I'm okay to leave that as a TODO if it's a bit too annoying right now.

v0d1ch replied:
I made it return `Maybe` type so we can handle it nicer.

noonio commented on hydra-cluster/test/Test/BlockfrostChainSpec.hs:
```diff
@@ -70,13 +70,14 @@ blockfrostProjectPath = "./../blockfrost-project.txt"
 spec :: Spec
 spec = around (onlyWithBlockfrostProjectFile . showLogsOnFailure "BlockfrostChainSpec") $ do
-  it "can open, close & fanout a Head using Blockfrost" $ \tracer -> do
+  xit "can open, close & fanout a Head using Blockfrost" $ \tracer -> do
```
Could we do a `pendingWith` or whatever, just so it's a bit easier to find why this is being disregarded for now? I find "xit" so subtle i never see it 😅

noonio commented on hydra-cluster/test/Test/CardanoNodeSpec.hs:
```diff
@@ -40,38 +41,44 @@ spec = do
-        doesFileExist (unFile nodeSocket) `shouldReturn` True
+      withCardanoNodeDevnet tr tmp $ \blockTime backend -> do
+        let nodeSocket' =
+              case Backend.getOptions backend of
```
Same as somewhere else; maybe we can do this check at the opening of the `do` block

noonio commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -365,32 +361,36 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+
+              let nodeSocket' = case Backend.getOptions backend of
+                    Direct DirectOptions{nodeSocket} -> nodeSocket
+                    _ -> error "Unexpected Blockfrost backend"
```
same

noonio commented on hydra-cluster/test/Test/Hydra/Cluster/HydraClientSpec.hs:
```diff
@@ -257,26 +257,30 @@ scenarioSetup tracer tmpDir action = do
+
+    let nodeSocket' = case Backend.getOptions backend of
+          Direct DirectOptions{nodeSocket} -> nodeSocket
+          _ -> error "Unexpected Blockfrost backend"
```
same

noonio reviewed (changes requested):
Very nice!

locallycompact commented:
How would we disable the blockfrost tests for merge requests after this merge, or is that done in this PR?

v0d1ch committed:
Make awaiting for tx utxo work

We used to call awaitUTxO in the blockfrost test separately. This should
be done when posting blockfrost transactions (like publishing scripts)
so we know that the following blockfrost call will have fresh UTxO to
work with.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Small refactor to not halt if db folder is not there.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch commented:
> https://github.com/cardano-scaling/hydra/pull/2057#issuecomment-2983763719 

@locallycompact now blockfrost is the option when running smoke-tests in CI so no need to run the one we have in the cluster. Eventually we should be able to swap backends so we can run all cluster tests using blockfrost if need be.

v0d1ch committed:
PR review comments

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ffakenz commented on .github/workflows/smoke-test.yaml:
```diff
@@ -60,7 +62,13 @@ jobs:
+
+        if [[ ${{secrets.blockfrost_token}} != '' ]]; then
+          echo "${{secrets.blockfrost_token}}" > blockfrost-project.txt
+          nix develop .#exes --command bash -c "hydra-cluster --${{inputs.network}} --state-directory ${state_dir} ${HYDRA_SCRIPTS_ARG} ${USE_MITHRIL_ARG}"
```
I would expect smoke-test to have an input flag to run using the --blockfrost option.

Why is that not the case?

v0d1ch replied:
I added another network related to blockfrost. After this is merged we should talk to blockfrost people and get preprod and mainnet keys too.

ffakenz commented on .github/workflows/smoke-test.yaml:
```diff
@@ -60,6 +62,12 @@ jobs:
         fi
+
+        if [[ ${{secrets.blockfrost_token}} != '' ]]; then
+          echo "${{secrets.blockfrost_token}}" > blockfrost-project.txt
```
there's a step above to "Set up secret faucet key".

maybe we should also add one to "Set up secret Blockfrost project"?

wdyt?

v0d1ch replied:
Yes, maybe we could. I opted for most simple thing.

ffakenz commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -45,22 +57,23 @@ run options =
       Nothing -> do
         when (useMithril == UseMithril) $ do
-          removeDirectoryRecursive $ workDir </> "db"
+          removeDirectoryRecursive (workDir </> "db") `catch` (\(_ :: SomeException) -> pure ())
```
why do we ignore the exception?

v0d1ch replied:
This was a simple _life improvement_ since the only way to get exception is if the folder is not there. But in any case mithril will download the db so this should not stop the test in case the folder was empty to start with.

ffakenz commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -23,17 +29,23 @@ run options =
   withTracer (Verbose "hydra-cluster") $ \tracer -> do
     traceWith tracer ClusterOptions{options}
     let fromCardanoNode = contramap FromCardanoNode tracer
+    let blockfrostNetworks = [BlockfrostPreview]
```
why do we only support blockfrost-preview?

v0d1ch replied:
That's the only key we have from blockfrost people for now.

ffakenz commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -214,6 +209,11 @@ withCardanoNodeOnKnownNetwork tracer stateDirectory knownNetwork action = do
     Preproduction -> "preprod"
     Mainnet -> "mainnet"
     Sanchonet -> "sanchonet"
+    -- NOTE: Here we map blockfrost networks to cardano ones since we expect to find actor keys
```
why not having a separate config folder specifically for blockfrost?

v0d1ch replied:
It would be just a duplication of actor keys and since blockfrost is just a proxy for cardano networks I didn't bother to copy keys

ffakenz commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -342,20 +336,22 @@ computeBlockTime slotLength activeSlotsCoeff =
-    eraHistory <- queryEraHistory networkId nodeSocket QueryTip
-    tipSlotNo <- queryTipSlotNo networkId nodeSocket
+    eraHistory <- Backend.queryEraHistory backend QueryTip
+    tipSlotNo <- fromMaybe 0 . Api.chainPointToSlotNo <$> Backend.queryTip backend
```
this pattern `fromMaybe 0 . Api.chainPointToSlotNo <$> Backend.queryTip backend` repeats a lot.

some dry would be nice :)

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -148,88 +143,85 @@ seedFromFaucetBlockfrost receivingVerificationKey lovelace = do
-          Nothing -> failure $ "Could not find script output: " <> decodeUtf8 (encodePretty newUtxo)
-          Just u -> pure u
- where
-  changeAddress = mkVkAddress networkId
```
I think having this variable makes the code more readable.

v0d1ch replied:
How? It is used at only one place?
